osu.Game/Rulesets/Replays/ReplayFrame.cs

```diff
@@ -30,5 +30,10 @@ public ReplayFrame(double time)
         {
             Time = time;
         }
+
+        /// <summary>
+        /// Whether this frame is equivalent to <paramref name="other"/> with respect to replay recording.
+        /// </summary>
+        public virtual bool IsEquivalentTo(ReplayFrame other) => Time == other.Time;
```

Comment:
I'm honestly not sure there would be a case where time based comparison alone isn't enough, ie I don't know the overloads are required. But seems fine either way.

Reply:
I have seen such cases yes. Think move input and press input in a single frame. Both will emit replay frames but the press input may not yet be visible on the frame emitted via the move input. 

Whether deduplication of *those* frames should be done is a good question but one I didn't particularly want to broach in this PR because that is the pre-existing state but I can consider doing something about it in either this or a separate PR if you think it's worth it.

Reply:
This might be a good time to ensure sanity in such cases – one frame per timestamp – while we're here correcting mistakes. I believe the last frame pushed with same timestamp should take precedence (overwrite others).

I'm fine with that happening in this PR or in a follow-up, your choice.

Reply:
I'll see what I can do. I will note at this time that the most likely problem here is *spectator* - while for local recording I absolutely can replace the last frame with the same timestamp, it's going to be rather difficult to *un-send* the frame from spectator server even with the frame bundle buffering that we have.

So it would probably have to be a solution oriented towards *delaying* sending the frame until we're certain that time has indeed advanced.

Reply:
Right. I wouldn't care if the duplication still happens in spectator flow, but I can understand if that lenience doesn't reduce complexity.

Reply:
If I'm allowed to ignore spectator flow then something like https://github.com/ppy/osu/pull/33148/commits/cd4ab9307c5ed7d160919aa265392ecbad29a138 should do the job.

Reply:
There's one caveat here, namely https://github.com/ppy/osu/pull/33148/commits/953a93309c341edf733c669d3df5f4a6aa801c2a - if it ever turns out that it's possible that a key press and release can occur in a single frame of actual play, then this may have unintended consequences. I don't have evidence that this can occur from my testing, but that also doesn't mean that it *can't* occur.

> since future reports will be either for `.osr` replays or spectator realtime.

Not quite true either, it's going to be local `.osr` replays vs spectator realtime *or* server-side `.osr` replays.

Reply:
Hmm, that's something I completely forgot. We might want some similar logic server-side as well (or in `FrameDataBundle`? assuming frames don't get split across two frame bundles)

Reply:
> or in `FrameDataBundle`? assuming frames don't get split across two frame bundles

It would definitely be easiest / most convenient to do there, but there's a possibility of splitting across bundles as you say. Maybe I do that for now and we'll see what happens?